### DISCUSSION TITLE
Deprecation notice: Class API option for new Quasar CLI projects

### DISCUSSION THREAD

**Discussion by IlCallo** [quasarframework/quasar#11204] [Announcements]
Hi everyone,
it's been quite some time since we tried to get in touch with https://github.com/vuejs/vue-class-component maintainer to discuss about a [dreaded bug](https://github.com/quasarframework/quasar/issues/9302) which prevents HMR when using Class API
The project unluckily is unmaintained since almost a year, stuck in an infinite beta phase

We received many requests from the community (and not only [our own](https://github.com/vuejs/vue-class-component/issues/569)) to put some kind of deprecation notice on the starter kit option, and we finally decided to take action

Class API support is now officially deprecated
**Existing projects using Class API will still work, the immediate effect of the deprecation will only affect the possibility to create new Class API projects**

This will be the deprecation plan:
- we'll give another try to solve the HMR problem
- we'll remove the Class API option from Quasar starter kit to prevent new projects to use it 
- Class API will still be supported for the duration of Quasar v2 lifecycle
- next year we'll evaluate if it makes sense to still support Class API in next major version (Quasar v3, expected in April 2022), mostly depending on what will happen to `vue-class-component` in next few months

We really do hope `vue-class-component` maintainer comes back, but up until then we cannot keep suggesting the community to use an unmaintained package

Let us know your feelings and thoughts on this, as we know many TS users into Quasar community still uses this API and we're interested in your opinion

**Comment by IlCallo** [comment]
Update: we were finally able to pin-point the problem into vue-loader code, a PR is on its way: https://github.com/vuejs/vue-loader/pull/1897
We'll now remove Class API option from the starter kit

If you're using Class API and you're happy we were able to solve this problem for the whole ecosystem, please consider [donating to the project](https://github.com/sponsors/rstoenescu) :)

**Reply by rstoenescu** [reply]
Seems that my PR has been merged and the vue-loader team has released a new version.
So upgraded to the new vue-loader in "@quasar/app" v3.1.9 -- highly recommended for those using the Class API.

Enjoy!

**Reply by ImBoop** [reply]
>  We'll now remove Class API option from the starter kit

Out of curiosity, why remove it if it'll still be supported for v2 and a major bug with it was fixed? The GH issue where someone requested the deprecation of vue-class-component remains opened with no major dissent in either direction.

In fact, class components is what made me switch to using Quasar in the first place. Please consider not doing this; asking us to sponsor as a thanks for fixing a bug for something you subsequently remove worries me that it'll be abandoned before Quasar 3

Much love <3

**Reply by IlCallo** [reply]
It's not a decision we took with light heart: we spent months trying to debug that HMR problem which resulted in a really bad DX for those using Class API
We reached for Vue core team and they could not help us, then tried to reach out to the Class API maintainer, to no avail since nobody heard a word from him since almost a year now.

We're not against Class API at all, but Quasar purpose is to raise the bar of web development, sponsor best practices and empower our devs to build something wonderful (and **maintenable**).
Suggesting an unmaintained package, from our point of view, means leading our devs to crash into a wall, more sooner than later.

> We really do hope vue-class-component maintainer comes back, but up until then we cannot keep suggesting the community to use an unmaintained package

As we wrote here, we're ready to re-enable it as soon as the Class API maintainer reach back to us, or if a reliable community-based fork comes up 

Rest assured: as for every deprecated feature in open source, **we won't drop Class API support for Quasar v2** (EOL near April 2023)
That said, if an alternative won't come up, we'll probably drop Class API support into Quasar v3

Don't take this answer as a full rejection: we're trying to be more open and transparent with the community, this announcement purpose was exactly to hear the feedback from you.
If anyone else have concerns, please join the discussion!

---

As I wrote up there, we disabled the Class API option into the starter kit, but the whole code to support it is still in place.
**!! At your own risk !!**, you can clone the starter kit repo locally, re-enable the Class API option and use it via `quasar create` `--kit` option to create a new Class API project

**Reply by ImBoop** [reply]
I understand. I'm biased as I absolutely love class components, so it hurts my (albeit tiny) heart to hear this though I do understand it. I guess my issue is more with the Vue team not saying one way or the other what the state of class components is.

We use quasar extensively because it's one of the few frameworks with sane TS support, and push come to shove we can just stay with v2.

Hopefully none of this will matter and vue will drop the radio silence with class components and we can live happily ever after in the end.

**Reply by Evertvdw** [reply]
@ImBoop I was kind of sad as well but I have started with the conversion to Composition API just now. It's not nearly as bad as I initially thought mainly because there is `<script setup lang="ts">` now and I'm using it together with `unplugin-auto-import` to make the code cleaner and overall its not that big of a switch from class components :)
